completions.go

```diff
@@ -135,23 +130,41 @@ func (c *Command) RegisterFlagCompletionFunc(flagName string, f func(cmd *Comman
 	if flag == nil {
 		return fmt.Errorf("RegisterFlagCompletionFunc: flag '%s' does not exist", flagName)
 	}
-	flagCompletionMutex.Lock()
-	defer flagCompletionMutex.Unlock()
+	// Ensure none of our relevant fields are nil.
+	c.initializeCompletionStorage()
 
-	if _, exists := flagCompletionFunctions[flag]; exists {
+	c.flagCompletionMutex.Lock()
+	defer c.flagCompletionMutex.Unlock()
+
+	// And attempt to bind the completion.
+	if _, exists := c.flagCompletionFunctions[flag]; exists {
 		return fmt.Errorf("RegisterFlagCompletionFunc: flag '%s' already registered", flagName)
 	}
-	flagCompletionFunctions[flag] = f
+	c.flagCompletionFunctions[flag] = f
 	return nil
 }
 
 // GetFlagCompletion returns the completion function for the given flag, if available.
-func GetFlagCompletion(flag *pflag.Flag) (func(cmd *Command, args []string, toComplete string) ([]string, ShellCompDirective), bool) {
-	flagCompletionMutex.RLock()
-	defer flagCompletionMutex.RUnlock()
+func (c *Command) GetFlagCompletion(flag *pflag.Flag) (func(cmd *Command, args []string, toComplete string) ([]string, ShellCompDirective), bool) {
+	c.initializeCompletionStorage()
+
+	c.flagCompletionMutex.RLock()
+	defer c.flagCompletionMutex.RUnlock()
+
+	completionFunc, exists := c.flagCompletionFunctions[flag]
+
+	// If found it here, return now
+	if completionFunc != nil && exists {
+		return completionFunc, exists
+	}
+
+	// If we are already at the root command level, return anyway
+	if !c.HasParent() {
+		return nil, false
+	}
 
-	completionFunc, exists := flagCompletionFunctions[flag]
-	return completionFunc, exists
+	// Or walk up the command tree.
+	return c.Parent().GetFlagCompletion(flag)
```

Comment:
3) Recursively walk up the command parents' tree until flagComp is found.
Behavior identical regardless of if the flagComp is searched with a flag pointer, or a flag name.